src/mpi/init/initthread.c

```diff
@@ -76,120 +76,140 @@ int MPI_Init_thread(int *argc, char ***argv, int required, int *provided)
     mpi_errno = MPIR_Datatype_init_predefined();
     MPIR_ERR_CHECK(mpi_errno);
 
-    /* ---- MPID_Init -------------------------------------------------- */
-    mpi_errno = MPID_Init();
-    MPIR_ERR_CHECK(mpi_errno);
+    if (MPIR_CVAR_DEBUG_HOLD) {
+        MPII_debugger_hold();
+    }
 
-    /* ---- MPII_Post_init --------------------------------------------- */
-    /* Commit pairtypes after device hooks are activated */
-    mpi_errno = MPIR_Datatype_commit_pairtypes();
-    MPIR_ERR_CHECK(mpi_errno);
 
-    /* Initialize collectives infrastructure */
-    mpi_errno = MPII_Coll_init();
+    /**********************************************************************/
+    /* Section 3: preinitialization is complete.  signal to error
+     * handling routines that core services are unavailable. */
+    /**********************************************************************/
+
+    MPL_atomic_store_int(&MPIR_Process.mpich_state, MPICH_MPI_STATE__PRE_INIT);
+
+
+    /**********************************************************************/
+    /* Section 4: contains the device initialization and setup and
+     * teardown needed for the device initialization. */
+    /**********************************************************************/
+
+    /* Setting isThreaded to 0 to trick any operations used within
+     * MPID_Init to think that we are running in a single threaded
+     * environment. */
+    MPIR_ThreadInfo.isThreaded = 0;
+
+    MPL_atomic_store_int(&MPIR_Process.mpich_state, MPICH_MPI_STATE__IN_INIT);
+
+    mpi_errno = MPID_Init(required, &MPIR_ThreadInfo.thread_provided);
     MPIR_ERR_CHECK(mpi_errno);
 
-    mpi_errno = MPII_post_init_global();
+
+    /**********************************************************************/
+    /* Section 5: contains post device initialization code.  Anything
+     * that we could not do before the device was initialized can be
+     * done now.  Try to keep this as small as possible.  Only things
+     * that *REALLY* cannot be done before device initialization
+     * should come here. */
+    /**********************************************************************/
+
+    /* pairtypes might need device hooks to be activated so the device
+     * can keep track of their creation.  that's why we need to do
+     * this after the device initialization.  */
+    mpi_errno = MPIR_Datatype_commit_pairtypes();
     MPIR_ERR_CHECK(mpi_errno);
 
+    /* FIXME: this is for rlog, and should be removed when we delete rlog */
+    MPII_Timer_init(MPIR_Process.comm_world->rank, MPIR_Process.comm_world->local_size);
+
     MPII_post_init_memory_tracing();
     MPII_init_dbg_logging();
-    /* if --enable-debuginfo is configured, prepare for debug info data.
-     * if MPIEXEC_DEBUG is set in environment, wait for debugger until MPIR_debug_gate is set to 1.
-     */
     MPII_Wait_for_debugger();
 
-    /* dup comm_self and creates progress thread (if needed) */
     mpi_errno = MPII_init_async();
```

Comment:
This will essentially launch a run-time progress call which will use CS protection and all the runtime assumptions. So it needs to be moved after setting `isThreaded` and probably have to be the last step anyway.

Reply:
Yes, good catch.  I'll fix it.